File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        if (s.length() <= 1) {
            return s.length();
        }
        char[] chars = s.toCharArray();
        int i = 0;
        int champSize = 0;
        int size = 0;
        Map<Character, Integer> map = new HashMap<>();
        // List<Integer> res = new ArrayList<>();
        while (i < chars.length) {
            if (!map.containsKey(chars[i])) {
                map.put(chars[i], i);
                size++;
                champSize = Math.max(champSize, size);
                i++;
            }
            else {
                champSize = Math.max(champSize, size);
                // res.add(size);
                int old = map.get(chars[i]);
                int gap = i - old;
                for (int j = i - size; j <= old; j++) {
                    map.remove(chars[j]);
                }
                map.put(chars[i], i);
                size = gap;
                i++;
            }
        }
        return champSize;
    }
}

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if s == "":
            return 0
        l,r = 0, 0
        champ = r-l+1
        d = {}
        chars = list(s)
        d[chars[l]] = l
        r += 1
        while l <= r and r < len(s):
            newChar = chars[r]
            if newChar not in d:
                champ = max(champ, r-l+1)
                d[newChar] = r
            else:
                repeatPos = d[newChar]
                if r-(repeatPos+1)+1 > repeatPos-l+1:
                    champ = max(champ, r-(repeatPos+1)+1)
                else:
                    champ = max(champ, repeatPos-l+1)
                for i in range(l, repeatPos+1):
                    d.pop(chars[i], None)
                d[newChar] = r
                l = repeatPos+1
            r += 1
        return champ
        